Auto-Type: Remember previous selected global match

hifi opened the pull request:
This makes using multi-stage login forms slightly easier as you can avoid typing the search terms multiple times. The last match is invalidated after 30 seconds which should be plenty if you are going through a login form.

Inspired by https://github.com/keepassxreboot/keepassxc/pull/5864#issuecomment-968916483 to avoid needing a complex sequence in the first place.

## Testing strategy
Manually on Linux.

## Type of change
- ✅ New feature (change that adds functionality)

michaelk83 commented on src/autotype/AutoTypeMatchView.h:
```diff
@@ -35,7 +35,7 @@ class AutoTypeMatchView : public QTableView
     AutoTypeMatch currentMatch();
     AutoTypeMatch matchFromIndex(const QModelIndex& index);
-    void setMatchList(const QList<AutoTypeMatch>& matches, bool selectFirst);
+    void setMatchList(const QList<AutoTypeMatch>& matches, int selectedIndex, bool selectFirst);
```
`selectFirst` looks redundant with `selectedIndex`.

hifi replied:
It's not actually redundant. `selectedIndex` is an index to the match list and not an index to the sorted view. I need to be able to distinguish from selecting the first _visible_ item in the list and the first item in the input list. If it can be easily compressed into a single argument that makes sense I'll gladly refactor it.

michaelk83 replied:
Then maybe just a rename to clarify the difference? Something like `matchIndex` or `inputIndex` or `selectedMatch` for `selectedIndex`, and `selectFirstVisible` for `selectFirst`?

(I'm still not entirely clear what `selectedIndex` actually does. May need to dig a little deeper in the code.)

hifi replied:
We discussed this on Matrix with @droidmonkey and decided to split this into three distinct functions where `setMatchList` just clears the selection so the call isn't too overloaded.

hifi replied:
I've now split it into separate functions.

michaelk83 commented on src/autotype/AutoTypeSelectDialog.cpp:
```diff
@@ -74,7 +75,7 @@ AutoTypeSelectDialog::AutoTypeSelectDialog(QWidget* parent)
         } else {
             // Reset to original match list
-            m_ui->view->setMatchList(m_matches, true);
+            m_ui->view->setMatchList(m_matches, -1, true);
```
Might want to use a constant for that `-1`. Though if `selectFirst` is removed, then this becomes 0 here.

hifi replied:
This should be fixed now with the refactor.

michaelk83 reviewed (commented):
This could use a doc update: how do you set up the Auto-Type sequences etc for multi-page login with this change?

droidmonkey commented on src/autotype/AutoType.cpp:
```diff
@@ -423,6 +425,11 @@ void AutoType::performGlobalAutoType(const QList<QSharedPointer<Database>>& dbLi
     }
 
+    // Invalidate last match if it's old enough
+    if (m_lastMatch.first != nullptr && QDateTime::currentMSecsSinceEpoch() - m_lastMatchTime > 30000) {
```
should be: `if (m_lastMatch.first && (QDateTime::currentMSecsSinceEpoch() - m_lastMatchTime) > 30000) {`

hifi replied:
As you please.

droidmonkey commented on src/autotype/AutoTypeMatchView.cpp:
```diff
@@ -78,7 +78,7 @@ void AutoTypeMatchView::keyPressEvent(QKeyEvent* event)
 }
 
-void AutoTypeMatchView::setMatchList(const QList<AutoTypeMatch>& matches, bool selectFirst)
+void AutoTypeMatchView::setMatchList(const QList<AutoTypeMatch>& matches, int selectedIndex, bool selectFirst)
```
Do we need the selectFirst parameter? Just default the selectedIndex param to 0.

hifi replied:
Yes unless there's a better way, see my comment for @michaelk83.

droidmonkey replied:
Recommend splitting these functions then. Add a function to select a specific (unsorted) index. Then remove the selectFirst parameter and revert back to the original behavior of selecting the top entry. Although technically that is `setCurrentIndex`

droidmonkey commented on src/autotype/AutoTypeSelectDialog.h:
```diff
@@ -63,6 +65,7 @@ private slots:
 
     QList<QSharedPointer<Database>> m_dbs;
     QList<AutoTypeMatch> m_matches;
+    AutoTypeMatch m_lastMatch;
```
This only seems to be used in 1 function, recommend not storing it.

hifi replied:
There are multiple call sites for `performSearch()` and it's always called from within the class so passing it as an argument wouldn't work, what would you suggest?

droidmonkey replied:
Code reading comprehension failure, I am adjusting my AI neural network now.

hifi commented:
@michaelk83 You don't need to setup anything special for Auto-Type sequences for this to work, it will just try its best to remember your last selection when you re-open the dialog with your global hotkey.

A mention of the flow in user guide would be good I presume.

droidmonkey commented on src/autotype/AutoType.cpp:
```diff
@@ -423,6 +425,11 @@ void AutoType::performGlobalAutoType(const QList<QSharedPointer<Database>>& dbLi
     }
 
+    // Invalidate last match if it's old enough
+    if (m_lastMatch.first && (QDateTime::currentMSecsSinceEpoch() - m_lastMatchTime) > 30000) {
```
Put the 30000 in a const int within anonymous namespace or a define. Also use Clock:: interface instead of QDateTime directly.

hifi replied:
Done, converted to using the seconds since we don't need more accuracy.

droidmonkey commented on src/autotype/AutoTypeMatchModel.h:
```diff
@@ -49,6 +49,7 @@ class AutoTypeMatchModel : public QAbstractTableModel
     QVariant headerData(int section, Qt::Orientation orientation, int role = Qt::DisplayRole) const override;
 
     void setMatchList(const QList<AutoTypeMatch>& matches);
+    QList<AutoTypeMatch>* matchList();
```
I don't like this at all, we should not be returning a pointer to an internal data member. This indicates an antipattern (also Qt does a shallow copy of container objects). Instead you should create a function (or use an existing) that searches the match list for the provided match and returns the index.

hifi replied:
Added a new one since we are fine with a match that matches only the entry but not the sequence.

droidmonkey commented on src/autotype/AutoTypeMatchView.cpp:
```diff
@@ -78,21 +79,43 @@ void AutoTypeMatchView::keyPressEvent(QKeyEvent* event)
+bool AutoTypeMatchView::selectMatch(const AutoTypeMatch& match)
+{
+    int selectedIndex = -1;
+    auto matches = m_model->matchList();
+    for (int i = 0; i < matches->size(); i++) {
+        const auto& currentMatch = matches->at(i);
+        if (currentMatch.first == match.first || (selectedIndex == -1 && currentMatch.first == match.first)) {
+            selectedIndex = i;
+        }
+    }
```
As mentioned above, this should be within the model itself since it controls the list of matches.

hifi replied:
Moved to model.

droidmonkey commented on src/autotype/AutoTypeSelectDialog.cpp:
```diff
@@ -74,7 +75,7 @@ AutoTypeSelectDialog::AutoTypeSelectDialog(QWidget* parent)
         } else {
             // Reset to original match list
-            m_ui->view->setMatchList(m_matches, true);
+            m_ui->view->setMatchList(m_matches);
```
Should we add a `selectFirstMatch` call here?

hifi replied:
I haven't seen any change in the behavior but for completeness why not.

hifi replied:
Added the call.

droidmonkey commented on src/autotype/AutoTypeMatchModel.cpp:
```diff
@@ -45,6 +45,20 @@ QModelIndex AutoTypeMatchModel::indexFromMatch(const AutoTypeMatch& match) const
+
+    for (int i = 0; i < m_matches.size(); i++) {
+        const auto& currentMatch = m_matches.at(i);
+        if (currentMatch.first == match.first || (row == -1 && currentMatch.first == match.first)) {
```
I don't understand this logic, the row == -1 doesn't make any difference since you are testing the same condition anyway.

phoerious commented on src/autotype/AutoTypeMatchModel.cpp:
```diff
@@ -45,6 +45,20 @@ QModelIndex AutoTypeMatchModel::indexFromMatch(const AutoTypeMatch& match) const
+    int row = -1;
+
+    for (int i = 0; i < m_matches.size(); i++) {
+        const auto& currentMatch = m_matches.at(i);
+        if (currentMatch.first == match.first || (row == -1 && currentMatch.first == match.first)) {
+            row = i;
+        }
```
You should add a ``break`` here to avoid unnecessary iterations and perhaps use STL iterators.

hifi committed:
Auto-Type: Remember previous selected global match

This makes using multi-stage login forms slightly easier as you
can avoid typing the search terms multiple times.

droidmonkey commented on src/autotype/AutoTypeMatchModel.cpp:
```diff
@@ -45,6 +45,24 @@ QModelIndex AutoTypeMatchModel::indexFromMatch(const AutoTypeMatch& match) const
+{
+    int row = -1;
+
+    for (int i = m_matches.size() - 1; i >= 0; i--) {
```
`--i`

droidmonkey commented on src/autotype/AutoTypeSelectDialog.cpp:
```diff
@@ -93,13 +86,23 @@ AutoTypeSelectDialog::~AutoTypeSelectDialog()
+    // disable changing search scope if we have no direct matches
+    m_ui->searchCheckBox->setDisabled(m_matches.isEmpty());
 
-    m_ui->view->setMatchList(m_matches, !m_matches.isEmpty() || !m_ui->search->text().isEmpty());
-    m_ui->searchCheckBox->setChecked(m_matches.isEmpty());
+    // changing check also performs search
+    if (m_ui->searchCheckBox->isChecked() != m_matches.isEmpty()) {
+        m_ui->searchCheckBox->setChecked(m_matches.isEmpty());
```
We use m_matches.isEmpty() three times, lets give it a variable name for ease of reading: `bool noMatches = m_matches.isEmpty()`

I also have to read this if statement several times to understand what is going on. I think we should just always set the search checkbox here to the noMatches boolean and get rid of the if statement. We could suppress the checkbox signals to prevent a double search.
